### DISCUSSION TITLE
Dynamic Route resolution using wildcards

### DISCUSSION THREAD

**Discussion by fritzchr** [tokio-rs/axum#2210] [Q&A]
I'm implementing a file service and designed the API in a way to resemble REST as much as possible. Due to this, i require two routes which take a path to a directory or file.

For example, the following routes should all be valid.
```
/dir/sub/sub2/file.txt -> Should be a file
/dir/sub/sub2/sub3/ -> Should be a dir
/dir/sub/sub2/something -> Should be a file
```
A '/' at the end of the path should be a directory, without the '/' at the end, the path should be recognized as a file. The Path can contain n amount of nested directories.

This is the equivalent HTTP Method. The query parameter changes whether it is a file or directory.
```
PUT https://<domain>/<directory>/?some_query=value
PUT https://<domain>/<directory>/<file>?some_other_query=value
```

I figured i could create a Route with Axum like that:
```
Router::new()
  .route("/*dir_path/", put(dir_handler))
  .route("/*dir_path/:filename", put(file_handler));
```

But a combiniation of these wildcards is not supported. Is there any way to achieve this 'dynamic' routing? Especially the different query parameters are important.

I am aware I could simply create different routes, e.g.
```
Router::new()
  .route("/dir/*dir_path", put(dir_handler))
  .route("/file/*file_path", put(file_handler));
```
But in my opinion this is not RESTful and I would like to avoid this solution if somehow possible.

**Comment by davidpdrsn** [comment]
I think creating different routes is fine. Not really clear to me why that's not RESTful but then again, REST means different things to different people 😅
